File: services/core-api/schemas/bikes.py
"""Bike domain schemas — CRUD, suspension spec, and summary stats."""

from __future__ import annotations

import uuid
from datetime import date, datetime
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class BikeStatus(str, Enum):
    """Current ownership / availability status."""

    owned = "owned"
    sold = "sold"
    stored = "stored"
    in_repair = "in_repair"


# ── Suspension spec ──


class SuspensionEndSettings(BaseModel):
    """Settings for one end of suspension (front fork or rear shock).

    Domain-specific names (compression_clicks, rebound_clicks, preload_turns,
    spring_rate_nmm, fork_height_mm) are canonical. The generic names used in
    the original schema (compression, rebound, preload, spring_rate) are accepted
    as input aliases for backwards compatibility.
    """

    model_config = ConfigDict(populate_by_name=True)

    brand: str | None = None

    # Damping — domain name is canonical, generic name accepted as alias
    compression_clicks: float | None = Field(None)
    rebound_clicks: float | None = Field(None)

    # Preload
    preload_turns: float | None = Field(None)

    # Spring
    spring_rate_nmm: float | None = Field(None)

    # Fork-specific geometry
    fork_height_mm: float | None = None
    oil_level: float | None = None
    ride_height: float | None = None

    @model_validator(mode="before")
    @classmethod
    def _accept_generic_aliases(cls, data: object) -> object:
        """Map legacy generic field names to domain-specific names on input."""
        if not isinstance(data, dict):
            return data
        d = dict(data)
        for generic, specific in (
            ("compression", "compression_clicks"),
            ("rebound", "rebound_clicks"),
            ("preload", "preload_turns"),
            ("spring_rate", "spring_rate_nmm"),
        ):
            if generic in d and specific not in d:
                d[specific] = d.pop(generic)
        return d


class SuspensionSpec(BaseModel):
    """Versioned suspension specification stored as JSONB. schema_version must be 1."""

    schema_version: int = Field(..., description="Must be 1")
    front: SuspensionEndSettings | None = None
    rear: SuspensionEndSettings | None = None

    @field_validator("schema_version")
    @classmethod
    def validate_schema_version(cls, v: int) -> int:
        if v != 1:
            raise ValueError("schema_version must be 1")
        return v


# ── Bike requests ──


class BikeCreate(BaseModel):
    """Add a new bike to the garage."""

    make: str = Field(..., min_length=1, max_length=100)
    model: str = Field(..., min_length=1, max_length=100)
    year: int | None = None
    vin: str | None = Field(None, max_length=17)
    color: str | None = Field(None, max_length=50)
    mileage_km: int | None = None
    engine_hours: float | None = None
    exhaust: str | None = None
    ecu: str | None = None
    gearing_front: int | None = None
    gearing_rear: int | None = None
    suspension_spec: SuspensionSpec | None = None
    notes: str | None = None
    status: BikeStatus = BikeStatus.owned


class BikeUpdate(BaseModel):
    """Partial update to bike fields. All fields optional."""

    make: str | None = Field(None, min_length=1, max_length=100)
    model: str | None = Field(None, min_length=1, max_length=100)
    year: int | None = None
    vin: str | None = Field(None, max_length=17)
    color: str | None = Field(None, max_length=50)
    mileage_km: int | None = None
    engine_hours: float | None = None
    exhaust: str | None = None
    ecu: str | None = None
    gearing_front: int | None = None
    gearing_rear: int | None = None
    suspension_spec: SuspensionSpec | None = None
    notes: str | None = None
    status: BikeStatus | None = None


# ── Bike responses ──


class BikeStats(BaseModel):
    """Computed summary statistics for a bike."""

    maintenance_count: int = 0
    last_maintenance_date: date | None = None
    modification_count: int = 0
    active_mods_count: int = 0
    session_count: int = 0
    best_lap_ms: int | None = None
    tire_pressure_last_checked: datetime | None = None


class BikeResponse(BaseModel):
    """Single bike with all fields."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    make: str
    model: str
    year: int | None = None
    vin: str | None = None
    color: str | None = None
    mileage_km: int | None = None
    engine_hours: float | None = None
    exhaust: str | None = None
    ecu: str | None = None
    gearing_front: int | None = None
    gearing_rear: int | None = None
    suspension_spec: SuspensionSpec
    notes: str | None = None
    status: BikeStatus
    deleted_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class BikeDetailResponse(BikeResponse):
    """Bike with summary stats (GET /garage/bikes/:id)."""

    stats: BikeStats = Field(default_factory=BikeStats)


class BikeListResponse(BaseModel):
    """List of bikes returned by GET /garage/bikes."""

    items: list[BikeResponse]

File: frontend/src/mocks/handlers/bikes.ts
import { http, HttpResponse } from 'msw';
import type { Bike, BikeDetail, CreateBikeRequest } from '@/api/types';

const BASE = '*/api/v1';

const MOCK_BIKES: Bike[] = [
  {
    id: 'bike-1',
    user_id: 'user-1',
    make: 'Ducati',
    model: 'Panigale V4 R',
    year: 2024,
    vin: 'ZDM14BKW0RB000123',
    color: 'Red',
    mileage_km: 4200,
    engine_hours: 82,
    exhaust: 'Akrapovic Full System',
    ecu: 'Ducati Performance',
    gearing_front: 16,
    gearing_rear: 42,
    suspension_spec: {
      schema_version: 1,
      front: {
        compression: 12,
        rebound: 14,
        preload: 8,
        spring_rate: 0.95,
        oil_level: 120,
        ride_height: null,
      },
      rear: {
        compression: 10,
        rebound: 12,
        preload: 6,
        spring_rate: 95,
        oil_level: null,
        ride_height: 5,
      },
    },
    notes: 'Track-prepped. Rearsets swapped to Woodcraft GP.',
    status: 'owned',
    deleted_at: null,
    created_at: '2024-03-15T10:00:00Z',
    updated_at: '2025-12-20T08:30:00Z',
  },
  {
    id: 'bike-2',
    user_id: 'user-1',
    make: 'Yamaha',
    model: 'YZF-R1',
    year: 2023,
    vin: null,
    color: 'Blue',
    mileage_km: 11200,
    engine_hours: null,
    exhaust: 'Graves Motorsports',
    ecu: null,
    gearing_front: 16,
    gearing_rear: 45,
    suspension_spec: {
      schema_version: 1,
      front: {
        compression: 8,
        rebound: 10,
        preload: 5,
      },
      rear: {
        compression: 6,
        rebound: 8,
        preload: 4,
      },
    },
    notes: null,
    status: 'owned',
    deleted_at: null,
    created_at: '2023-06-01T12:00:00Z',
    updated_at: '2025-11-10T14:00:00Z',
  },
];

const MOCK_BIKE_DETAILS: Record<string, BikeDetail> = {
  'bike-1': {
    ...MOCK_BIKES[0],
    stats: {
      maintenance_count: 12,
      modification_count: 8,
      session_count: 24,
      best_lap_ms: 98432,
    },
  },
  'bike-2': {
    ...MOCK_BIKES[1],
    stats: {
      maintenance_count: 5,
      modification_count: 3,
      session_count: 6,
      best_lap_ms: 105200,
    },
  },
};

export const bikeHandlers = [
  // GET /api/v1/garage/bikes
  http.get(`${BASE}/garage/bikes`, () => {
    return HttpResponse.json(MOCK_BIKES);
  }),

  // POST /api/v1/garage/bikes
  http.post(`${BASE}/garage/bikes`, async ({ request }) => {
    const body = (await request.json()) as CreateBikeRequest;
    const newBike: Bike = {
      id: `bike-${Date.now()}`,
      user_id: 'user-1',
      make: body.make,
      model: body.model,
      year: body.year ?? null,
      vin: body.vin ?? null,
      color: body.color ?? null,
      mileage_km: body.mileage_km ?? null,
      engine_hours: body.engine_hours ?? null,
      exhaust: body.exhaust ?? null,
      ecu: body.ecu ?? null,
      gearing_front: body.gearing_front ?? null,
      gearing_rear: body.gearing_rear ?? null,
      suspension_spec: body.suspension_spec ?? { schema_version: 1 },
      notes: body.notes ?? null,
      status: body.status ?? 'owned',
      deleted_at: null,
      created_at: new Date().toISOString(),
      updated_at: new Date().toISOString(),
    };
    MOCK_BIKES.push(newBike);
    MOCK_BIKE_DETAILS[newBike.id] = {
      ...newBike,
      stats: {
        maintenance_count: 0,
        modification_count: 0,
        session_count: 0,
        best_lap_ms: null,
      },
    };
    return HttpResponse.json(newBike, { status: 201 });
  }),

  // GET /api/v1/garage/bikes/:id
  http.get(`${BASE}/garage/bikes/:id`, ({ params }) => {
    const detail = MOCK_BIKE_DETAILS[params.id as string];
    if (!detail) {
      return HttpResponse.json(
        { error: 'Bike not found', code: 'NOT_FOUND' },
        { status: 404 },
      );
    }
    return HttpResponse.json(detail);
  }),

  // PATCH /api/v1/garage/bikes/:id
  http.patch(`${BASE}/garage/bikes/:id`, async ({ params, request }) => {
    const detail = MOCK_BIKE_DETAILS[params.id as string];
    if (!detail) {
      return HttpResponse.json(
        { error: 'Bike not found', code: 'NOT_FOUND' },
        { status: 404 },
      );
    }
    const body = (await request.json()) as Record<string, unknown>;
    const updated: Bike = {
      ...detail,
      ...body,
      updated_at: new Date().toISOString(),
    } as Bike;
    return HttpResponse.json(updated);
  }),

  // DELETE /api/v1/garage/bikes/:id
  http.delete(`${BASE}/garage/bikes/:id`, ({ params }) => {
    const detail = MOCK_BIKE_DETAILS[params.id as string];
    if (!detail) {
      return HttpResponse.json(
        { error: 'Bike not found', code: 'NOT_FOUND' },
        { status: 404 },
      );
    }
    return new HttpResponse(null, { status: 204 });
  }),
];
